tools/chainsync-loadtest/src/concurrency/once.rs

```diff
@@ -0,0 +1,58 @@
+use std::future::Future;
+use std::sync::RwLock;
+
+fn is_send<T: Send>() {}
+fn is_sync<T: Sync>() {}
+
+#[allow(dead_code)]
+fn test<T: Clone + Send + Sync>() {
+    is_send::<Once<T>>();
+    is_sync::<Once<T>>();
+}
+
+// Once is a synchronization primitive, which stores a single value.
+// This value can be set at most once, and multiple consumers are
+// allowed to wait for that value.
+pub struct Once<T: Clone + Send + Sync> {
+    value: RwLock<Option<T>>,
+    notify: tokio::sync::Notify,
```

Comment:
> Mixing in a synchronous RwLock into what otherwise seems like a type for use in async contexts seems like a potential footgun.

Synchronization of asynchronous tasks requires the use of synchronous primitives at some layer of abstraction. It is as footgun'ish as using mutexes in general.

I missed the fact that std::sync::RwLock is not guaranteed to be fair. How about I just replace it with parking_lot::RwLock ? It should be robust enough, given that by design there should be just 1 write, so there is no point in optimizing for multiple writers (here we just care about correctness).

Reply:
Yes, however tokio-provided async primitives (which internally are using sync primitives anyway) didn't satisfy my requirements, therefore I've augmented them to construct another async primitive (also by using some sync primitives internally).